watertap/unit_models/zero_order/coag_and_floc_zo.py

```diff
@@ -0,0 +1,226 @@
+
+        def rule_chem_flow(blk, t, j):
+            if j == 'alum':
+                chemical_dosage = blk.alum_dose[t]
+            elif j == 'polymer':
+                chemical_dosage = blk.polymer_dose[t]
+            return (blk.chemical_flow_mass[t, j] ==
+                    pyunits.convert(chemical_dosage * blk.properties[t].flow_vol,
+                                    to_units=pyunits.kg/pyunits.s))
+        self.chemical_flow_constraint = Constraint(self.flowsheet().time,
+                                                   ["alum", "polymer"],
+                                                   rule=rule_chem_flow)
+
+        def rule_anionic_polymer_dose(blk, t):
+            return (blk.anion_to_cation_polymer_ratio[t] * blk.polymer_dose[t]
+                    / (blk.anion_to_cation_polymer_ratio[t] + 1))
+        self.anionic_polymer_dose = Expression(self.flowsheet().config.time,
+                                               rule=rule_anionic_polymer_dose)
+
+        def rule_cationic_polymer_dose(blk, t):
+            return (blk.polymer_dose[t]
+                    / (blk.anion_to_cation_polymer_ratio[t] + 1))
+        self.cationic_polymer_dose = Expression(self.flowsheet().config.time,
+                                                rule=rule_cationic_polymer_dose)
+
+        self._perf_expr_dict["Anionic Polymer Dosage (mg/L)"] = self.anionic_polymer_dose
+        self._perf_expr_dict["Cationic Polymer Dosage (mg/L)"] = self.cationic_polymer_dose
+
+        # TODO: WT3 doesn't include pump electricity. Consider adding after case study validation
+        # pump_electricity(self, self.chemical_flow_vol)
+
+        @self.Constraint(self.flowsheet().time,
+                         doc='Constraint for rapid mix power consumption')
+        def rule_power_rapid_mix(b, t):
+            return (b.power_rapid_mix[t] ==
+                    pyunits.convert(b.num_rapid_mixers
+                    * b.properties[t].visc_d
+                    * b.rapid_mix_basin_vol
+                    * b.G_rapid_mix[t] ** 2, to_units=pyunits.kW))
+
+        @self.Constraint(self.flowsheet().time,
+                         doc='Constraint for floc power consumption')
+        def rule_power_floc(b, t):
+            return (b.power_floc[t] ==
+                    pyunits.convert(b.num_floc_mixers
+                    * b.properties[t].visc_d
+                    * b.floc_basin_vol
+                    * b.G_floc[t] ** 2, to_units=pyunits.kW))
+
+        @self.Expression(self.flowsheet().time,
+                         doc='Total power consumption')
+        def total_power(b, t):
+            return b.power_floc[t] + b.power_rapid_mix[t]
+
+        self._perf_expr_dict["Total Power Consumption (kW)"] = self.total_power
+
+        @self.Constraint(self.flowsheet().time,
+                         doc='Electricity intensity with respect to inlet flowrate of unit')
+        def rule_energy_electric_flow_vol_inlet(b, t):
```

Comment:
Do we actually need this here, seeing as we are calculating consumption separately?

Reply:
I asked myself the same question, but the WT3 method for computing energy consumption returns the "electricity intensity". So I just kept consistent with the original implementation in case it is needed for the case studies for any reason.

Reply:
I see a few things here:

1. Electricity consumption is probably the more important metric and the one that will be used at the flowsheet level - intensity is inherently specific to the unit and thus not so useful at the flowsheet level (you can calculate a separate flowsheet intensity if needed, but it will not be the sum across the units).
2. In many cases, WT3 ends up doing something like `intensity = some_expression/Q` followed by `electricity = intensity*Q` - i.e. they should have just calculated the electricity demand straight up and forgotten about intensity (or at least calculated it from demand instead of the other way around).
3. Base on the above, I don't see much of a need for calculating intensity in these cases - personally I would only include it in the cases where it is a parameter used to calculate electricity demand.

Reply:
Fair enough- I'll remove it.